File: hub/api/geoip/geoip.go
// Package geoip wraps the MaxMind GeoLite2 database to enrich IP addresses
// with country, city, and ASN information.
//
// # Setup
//
// Download the free GeoLite2 databases from MaxMind:
//
//	curl -o /etc/netscope/GeoLite2-City.mmdb    https://… (requires free account)
//	curl -o /etc/netscope/GeoLite2-ASN.mmdb     https://…
//
// Set the paths via environment variables:
//
//	GEOIP_CITY_DB=/etc/netscope/GeoLite2-City.mmdb
//	GEOIP_ASN_DB=/etc/netscope/GeoLite2-ASN.mmdb
//
// If the files are absent the Reader degrades gracefully — all lookups return
// empty GeoInfo structs, so ingest continues without enrichment.
package geoip

import (
	"log/slog"
	"net"
	"sync"

	"github.com/oschwald/geoip2-golang"
)

// GeoInfo holds the enrichment data for a single IP address.
type GeoInfo struct {
	// CountryCode is the ISO 3166-1 alpha-2 country code (e.g. "US", "DE").
	CountryCode string `json:"country_code"`
	// CountryName is the English country name.
	CountryName string `json:"country_name"`
	// City name (may be empty for small or rural IPs).
	City string `json:"city,omitempty"`
	// Latitude and Longitude of the best-guess location.
	Latitude  float64 `json:"latitude,omitempty"`
	Longitude float64 `json:"longitude,omitempty"`
	// ASNumber is the Autonomous System number (e.g. 15169 for Google).
	ASNumber uint `json:"as_number,omitempty"`
	// ASOrg is the registered organisation name for the AS.
	ASOrg string `json:"as_org,omitempty"`
}

// Reader is a thread-safe GeoIP enrichment client.
type Reader struct {
	cityDB *geoip2.Reader
	asnDB  *geoip2.Reader
	mu     sync.RWMutex
}

// New opens the GeoLite2 City and ASN databases at the given paths.
// Pass empty strings to skip loading the respective database.
// The returned Reader is always valid — missing databases are simply skipped.
func New(cityDBPath, asnDBPath string) *Reader {
	r := &Reader{}

	if cityDBPath != "" {
		db, err := geoip2.Open(cityDBPath)
		if err != nil {
			slog.Warn("geoip: city database unavailable", "path", cityDBPath, "err", err)
		} else {
			r.cityDB = db
			slog.Info("geoip: city database loaded", "path", cityDBPath)
		}
	}

	if asnDBPath != "" {
		db, err := geoip2.Open(asnDBPath)
		if err != nil {
			slog.Warn("geoip: ASN database unavailable", "path", asnDBPath, "err", err)
		} else {
			r.asnDB = db
			slog.Info("geoip: ASN database loaded", "path", asnDBPath)
		}
	}

	return r
}

// Lookup returns geo enrichment for the given IP address string.
// Returns an empty GeoInfo (not an error) when the databases are unavailable
// or the IP is private/loopback.
func (r *Reader) Lookup(ipStr string) GeoInfo {
	ip := net.ParseIP(ipStr)
	if ip == nil || isPrivate(ip) {
		return GeoInfo{}
	}

	r.mu.RLock()
	defer r.mu.RUnlock()

	info := GeoInfo{}

	if r.cityDB != nil {
		if record, err := r.cityDB.City(ip); err == nil {
			info.CountryCode = record.Country.IsoCode
			if name, ok := record.Country.Names["en"]; ok {
				info.CountryName = name
			}
			if len(record.City.Names) > 0 {
				info.City = record.City.Names["en"]
			}
			info.Latitude = record.Location.Latitude
			info.Longitude = record.Location.Longitude
		}
	}

	if r.asnDB != nil {
		if record, err := r.asnDB.ASN(ip); err == nil {
			info.ASNumber = record.AutonomousSystemNumber
			info.ASOrg = record.AutonomousSystemOrganization
		}
	}

	return info
}

// Close releases the database file handles.
func (r *Reader) Close() {
	r.mu.Lock()
	defer r.mu.Unlock()
	if r.cityDB != nil {
		r.cityDB.Close()
	}
	if r.asnDB != nil {
		r.asnDB.Close()
	}
}

// ── helpers ───────────────────────────────────────────────────────────────────

var privateRanges []*net.IPNet

func init() {
	cidrs := []string{
		"10.0.0.0/8", "172.16.0.0/12", "192.168.0.0/16",
		"127.0.0.0/8", "169.254.0.0/16",
		"::1/128", "fc00::/7", "fe80::/10",
	}
	for _, cidr := range cidrs {
		_, n, _ := net.ParseCIDR(cidr)
		if n != nil {
			privateRanges = append(privateRanges, n)
		}
	}
}

func isPrivate(ip net.IP) bool {
	for _, n := range privateRanges {
		if n.Contains(ip) {
			return true
		}
	}
	return false
}

File: desktop/src-tauri/src/geoip.rs
/// GeoIP enrichment using MaxMind GeoLite2 databases.
///
/// Databases are looked up from `~/.netscope/` by default.
/// Download free copies from https://dev.maxmind.com/geoip/geolite2-free-geolocation-data
use maxminddb::{geoip2, MaxMindDBError, Reader};
use std::net::IpAddr;
use std::path::{Path, PathBuf};

pub struct GeoIpReader {
    city: Option<Reader<Vec<u8>>>,
    asn: Option<Reader<Vec<u8>>>,
}

#[derive(Debug, Clone)]
pub struct GeoInfo {
    pub country_code: String,
    pub country_name: String,
    pub city: String,
    pub asn: u32,
    pub as_org: String,
}

fn is_private(ip: &IpAddr) -> bool {
    match ip {
        IpAddr::V4(v4) => {
            v4.is_private()
                || v4.is_loopback()
                || v4.is_link_local()
                || v4.is_broadcast()
                || v4.is_unspecified()
        }
        IpAddr::V6(v6) => v6.is_loopback() || v6.is_unspecified(),
    }
}

impl GeoIpReader {
    /// Open databases at the given paths. Either path may be absent — we degrade
    /// gracefully to whichever DB is available.
    pub fn open(city_path: &Path, asn_path: &Path) -> Result<Self, String> {
        let city = if city_path.exists() {
            Some(
                Reader::open_readfile(city_path)
                    .map_err(|e| format!("Failed to open GeoLite2-City: {e}"))?,
            )
        } else {
            None
        };

        let asn = if asn_path.exists() {
            Some(
                Reader::open_readfile(asn_path)
                    .map_err(|e| format!("Failed to open GeoLite2-ASN: {e}"))?,
            )
        } else {
            None
        };

        if city.is_none() && asn.is_none() {
            return Err("No GeoIP databases found".to_string());
        }

        Ok(GeoIpReader { city, asn })
    }

    /// Attempt to load databases from the default location (`~/.netscope/`).
    pub fn try_default() -> Option<Self> {
        let base = default_db_dir()?;
        Self::open(
            &base.join("GeoLite2-City.mmdb"),
            &base.join("GeoLite2-ASN.mmdb"),
        )
        .ok()
    }

    /// Resolve geographic info for a single IP string. Returns `None` for
    /// private/loopback addresses or any parse/lookup failure.
    pub fn lookup(&self, ip_str: &str) -> Option<GeoInfo> {
        let ip: IpAddr = ip_str.parse().ok()?;
        if is_private(&ip) {
            return None;
        }

        // ── City / country ────────────────────────────────────────────────────
        let (country_code, country_name, city_name) = match &self.city {
            Some(city_db) => match city_db.lookup::<geoip2::City>(ip) {
                Ok(rec) => {
                    let cc = rec
                        .country
                        .as_ref()
                        .and_then(|c| c.iso_code)
                        .unwrap_or("??")
                        .to_string();
                    let name = rec
                        .country
                        .as_ref()
                        .and_then(|c| c.names.as_ref())
                        .and_then(|n| n.get("en").copied())
                        .unwrap_or("Unknown")
                        .to_string();
                    let city = rec
                        .city
                        .as_ref()
                        .and_then(|c| c.names.as_ref())
                        .and_then(|n| n.get("en").copied())
                        .unwrap_or("")
                        .to_string();
                    (cc, name, city)
                }
                Err(MaxMindDBError::AddressNotFoundError(_)) => {
                    ("??".to_string(), "Unknown".to_string(), String::new())
                }
                Err(_) => return None,
            },
            None => ("??".to_string(), "Unknown".to_string(), String::new()),
        };

        // ── ASN ───────────────────────────────────────────────────────────────
        let (asn, as_org) = match &self.asn {
            Some(asn_db) => match asn_db.lookup::<geoip2::Asn>(ip) {
                Ok(rec) => (
                    rec.autonomous_system_number.unwrap_or(0),
                    rec.autonomous_system_organization
                        .unwrap_or("")
                        .to_string(),
                ),
                Err(_) => (0, String::new()),
            },
            None => (0, String::new()),
        };

        Some(GeoInfo {
            country_code,
            country_name,
            city: city_name,
            asn,
            as_org,
        })
    }
}

/// `~/.netscope/`
pub fn default_db_dir() -> Option<PathBuf> {
    dirs::home_dir().map(|h| h.join(".netscope"))
}
